frontend/model/contracts/chatroom.js

```diff
@@ -124,17 +124,24 @@ function createNotificationData (
 }
 
 function emitMessageEvent ({ type, contractID, hash, state }: {
-  type: string, contractID: string, hash: string, state: Object
+  type: string,
+  contractID: string,
+  hash: string,
+  state?: Object
 }): void {
-  for (let i = state.messages.length - 1; i >= 0; i--) {
-    if (state.messages[i].id === hash) {
-      sbp('okTurtles.events/emit', `${CHATROOM_MESSAGE_ACTION}-${contractID}`, {
-        type,
-        data: { message: state.messages[i] }
-      })
-      break
+  let data = {}
+  if (type === MESSAGE_ACTION_TYPES.ADD_MESSAGE || type === MESSAGE_ACTION_TYPES.EDIT_MESSAGE) {
+    const messagesCount = state?.messages.length || 0
+    for (let i = messagesCount - 1; i >= 0; i--) {
```

Comment:
As you know, the purpose is to find a `message` having same hash as it's `id`.
I think if we do reverse loop, we can save calculation and time to get the right message.
That's because in most cases, we will update message which we sent just before.
We don't update message which I sent long time ago.

Reply:
Fair point. I've also been thinking that keeping track of the `id` <-> `message` mappings could sometimes allow to do plain lookups such as `messagesById[hash]` instead of loops, but the mappings object would have to be updated every time a message is added or removed, so I'm not sure.

Reply:
In `chat-contract-optimize` branch, I am going to remove `state.messages` to save in the vuex store.
So we can save disk and performance too.